### PR TITLE
Multi domain select

### PR DESCRIPTION
**Pull request by rafalpospiech**
chrome_2019-05-16_10-49-08

### CODE REVIEW

File: layouts/basic/modules/Vtiger/Edit/Field/MultiDomain.tpl [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 3.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
+{strip}
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/Edit/Field/MultiDomain.tpl]: `<!-- tpl-Base-Edit-Field-MultiDomain -->`

File: layouts/basic/modules/Vtiger/Edit/Field/MultiDomain.tpl [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+	{assign var="SPECIAL_VALIDATOR" value=$FIELD_MODEL->getValidator()}
+	{assign var=FIELD_VALUE value=$FIELD_MODEL->getEditViewDisplayValue($FIELD_MODEL->get('fieldvalue'),$RECORD)}
+<div class="tpl-Edit-Field-MultiDomain">
+	<input type="hidden" name="{$FIELD_MODEL->getFieldName()}" value="" />
+	<select id="{$MODULE}_{$VIEW}_fieldName_{$FIELD_MODEL->getName()}" title="{\App\Language::translate($FIELD_MODEL->getFieldLabel(), $MODULE)}" multiple data-tags="true" class="js-multi-domain select2 form-control col-md-12" name="{$FIELD_MODEL->getFieldName()}[]" data-fieldinfo='{$FIELD_INFO}' {if $FIELD_MODEL->isMandatory() eq true} data-validation-engine="validate[required,funcCall[Vtiger_Base_Validator_Js.invokeValidation]]" {else} data-validation-engine="validate[funcCall[Vtiger_Base_Validator_Js.invokeValidation]]"{/if} data-validator={\App\Json::encode([['name'=>'MultiDomain']])} {if $FIELD_MODEL->isEditableReadOnly()}readonly="readonly"{/if}>
+		{foreach item=PICKLIST_VALUE from=$FIELD_VALUE}
+			<option value="{\App\Purifier::encodeHtml($PICKLIST_VALUE)}" {if $PICKLIST_VALUE}selected{/if}>{$PICKLIST_VALUE}</option>
+		{/foreach}
+	</select>
+</div>
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/Edit/Field/MultiDomain.tpl]: `<!-- /tpl-Base-Edit-Field-MultiDomain -->`

File: modules/OSSMail/models/Mail.php [hunk 3]
```diff
@@ -253,65 +252,115 @@ public function getEmail($name)
...
+	 */
+	public function searchByEmails(array $emails, string $moduleName, string $fieldName)
+	{
+		$return = [];
+		foreach ($emails as $email) {
+			if (empty($email)) {
+				continue;
+			}
+			$name = 'MSFindEmail_' . $moduleName . '_' . $fieldName;
+			$cache = App\Cache::get($name, $email);
```
Comment (mariuszkrzaczkowski) [inline modules/OSSMail/models/Mail.php]:
```suggestion
			$cache = App\Cache::staticGet($name, $email);
```

File: modules/OSSMail/models/Mail.php [hunk 4]
```diff
@@ -253,65 +252,115 @@ public function getEmail($name)
...
+			if ($cache !== false) {
```
Comment (mariuszkrzaczkowski) [inline modules/OSSMail/models/Mail.php]:
```suggestion
			if (App\Cache::staticHas($name, $email)) {
```

File: modules/OSSMail/models/Mail.php [hunk 5]
```diff
@@ -253,65 +252,115 @@ public function getEmail($name)
...
+					while (($crmid = $dataReader->readColumn(0)) !== false) {
+						$ids[] = $crmid;
+					}
+					$dataReader->close();
+					$return = array_merge($return, $ids);
+				}
+				if (empty($ids)) {
+					$ids = 0;
+				}
+				App\Cache::save($name, $email, $ids);
```
Comment (mariuszkrzaczkowski) [inline modules/OSSMail/models/Mail.php]:
```suggestion
				App\Cache::staticSave($name, $email, $ids);
```

File: modules/OSSMail/models/Mail.php [hunk 6]
```diff
@@ -253,65 +252,115 @@ public function getEmail($name)
...
+	 * @return array crmids
+	 */
+	public function searchByEmails(array $emails, string $moduleName, string $fieldName)
+	{
+		$return = [];
+		foreach ($emails as $email) {
+			if (empty($email)) {
+				continue;
+			}
+			$name = 'MSFindEmail_' . $moduleName . '_' . $fieldName;
```
Comment (mariuszkrzaczkowski) [inline modules/OSSMail/models/Mail.php]:
```suggestion
			$name = 'MailSearchByEmails' . $moduleName . '_' . $fieldName;
```

File: modules/OSSMail/models/Mail.php [hunk 7]
```diff
@@ -253,65 +252,115 @@ public function getEmail($name)
...
+		foreach ($emails as $email) {
+			$domain = mb_strtolower(explode('@', $email)[1]);
+			$cache = App\Cache::get($name, $domain);
+			if ($cache !== false) {
+				if ($cache != 0) {
+					$crmids = array_merge($crmids, $cache);
+				}
+			} else {
+				$crmids = App\Fields\MultiDomain::getCrmIds($domain, $fieldModel);
+				App\Cache::save($name, $domain, $crmids);
```
Comment (mariuszkrzaczkowski) [inline modules/OSSMail/models/Mail.php]:
```suggestion
				App\Cache::staticSave($name, $domain, $crmids);
```

File: modules/OSSMail/models/Mail.php [hunk 8]
```diff
@@ -253,65 +252,115 @@ public function getEmail($name)
...
+	 */
+	public function searchByDomains(Vtiger_Field_Model $fieldModel, array $emails)
+	{
+		$fieldName = $fieldModel->getName();
+		$moduleName = $fieldModel->getModuleName();
+		$name = 'MSFindEmail_' . $moduleName . '_' . $fieldName;
+		$crmids = [];
+		foreach ($emails as $email) {
+			$domain = mb_strtolower(explode('@', $email)[1]);
+			$cache = App\Cache::get($name, $domain);
```
Comment (mariuszkrzaczkowski) [inline modules/OSSMail/models/Mail.php]: ?

File: modules/OSSMail/models/Mail.php [hunk 9]
```diff
@@ -253,65 +252,115 @@ public function getEmail($name)
...
+			if ($cache !== false) {
```
Comment (mariuszkrzaczkowski) [inline modules/OSSMail/models/Mail.php]: ?

File: modules/OSSMail/models/Mail.php [hunk 10]
```diff
@@ -253,65 +252,115 @@ public function getEmail($name)
...
+	 * @param Vtiger_Field_Model $fieldModel
+	 * @param string[]           $emails
+	 *
+	 * @return array crmids
+	 */
+	public function searchByDomains(Vtiger_Field_Model $fieldModel, array $emails)
+	{
+		$fieldName = $fieldModel->getName();
+		$moduleName = $fieldModel->getModuleName();
+		$name = 'MSFindEmail_' . $moduleName . '_' . $fieldName;
```
Comment (mariuszkrzaczkowski) [inline modules/OSSMail/models/Mail.php]:
```suggestion
		$name = 'MailSearchByDomains' . $moduleName . '_' . $fieldName;
```

File: modules/OSSMail/models/Mail.php [hunk 11]
```diff
@@ -253,65 +252,115 @@ public function getEmail($name)
...
 		} else {
 			$emails = (array) $emails;
 		}
+		$emailSearchList = OSSMailScanner_Record_Model::getEmailSearchList();
 		if (!empty($emailSearchList)) {
 			foreach ($emailSearchList as $field) {
 				$enableFind = true;
 				$row = explode('=', $field);
 				$moduleName = $row[1];
+				$fieldModel = Vtiger_Field_Model::getInstance($row[0], Vtiger_Module_Model::getInstance($moduleName));
```
Comment (mariuszkrzaczkowski) [inline modules/OSSMail/models/Mail.php]: Vtiger_Module_Model::getInstance($moduleName)  
before the loop

File: modules/OSSMail/models/Mail.php [hunk 12]
```diff
@@ -253,65 +252,115 @@ public function getEmail($name)
...
-								}
-								$dataReader->close();
-								$return = array_merge($return, $ids);
-							}
-							if (empty($ids)) {
-								$ids = 0;
-							}
-							Vtiger_Cache::set($name, $email, $ids);
-						}
+					if ($fieldModel->get('uitype') === 319) {
```
Comment (mariuszkrzaczkowski) [inline modules/OSSMail/models/Mail.php]:
```suggestion
					if ($fieldModel->getUIType() === 319) {
```

File: modules/Vtiger/models/Record.php [hunk 13]
```diff
@@ -539,6 +539,7 @@ public function getValuesForSave()
 		foreach ($saveFields as &$fieldName) {
 			$fieldModel = $moduleModel->getFieldByName($fieldName);
 			if ($fieldModel) {
+				$fieldParams = $fieldModel->getFieldParams();
```
Comment (rskrzypczak) [inline modules/Vtiger/models/Record.php]: ?

File: modules/OSSMailScanner/models/Record.php [hunk 14]
```diff
@@ -428,7 +428,7 @@ public static function getEmailSearch($module = false)
 	 */
 	public static function getEmailSearchList()
 	{
-		$cache = Vtiger_Cache::get('Mail', 'EmailSearchList');
+		$cache = \App\Cache::get('Mail', 'EmailSearchList');
```
Comment (mariuszkrzaczkowski) [inline modules/OSSMailScanner/models/Record.php]: static

File: modules/OSSMailScanner/models/Record.php [hunk 15]
```diff
@@ -439,7 +439,7 @@ public static function getEmailSearchList()
 		if (!empty($value)) {
 			$return = explode(',', $value);
 		}
-		Vtiger_Cache::set('Mail', 'EmailSearchList', $return);
+		\App\Cache::save('Mail', 'EmailSearchList', $return);
```
Comment (mariuszkrzaczkowski) [inline modules/OSSMailScanner/models/Record.php]: static

File: modules/Vtiger/uitypes/MultiDomain.php [hunk 16]
```diff
@@ -0,0 +1,131 @@
+<?php
+
+/*
+ * MultiDomain class
+ *
+ * @copyright YetiForce Sp. z o.o.
+ * @license YetiForce Public License 3.0 (licenses/LicenseEN.txt or yetiforce.com)
+ * @author Rafal Pospiech <r.pospiech@yetiforce.com>
+*/
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/uitypes/MultiDomain.php]: format

File: modules/Vtiger/uitypes/MultiDomain.php [hunk 17]
```diff
@@ -0,0 +1,131 @@
+<?php
+
+/*
+ * MultiDomain class
+ *
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/uitypes/MultiDomain.php]: @package   UIType

File: modules/Vtiger/uitypes/MultiDomain.php [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+		$fieldModel = $this->getFieldModel();
+		$fieldParams = $fieldModel->getFieldParams();
+		foreach ($value as $item) {
+			if (!is_string($item)) {
+				throw new \App\Exceptions\Security('ERR_ILLEGAL_FIELD_VALUE||' . $this->getFieldModel()->getFieldName() . '||' . $this->getFieldModel()->getModuleName() . '||' . $item, 406);
+			}
+			if ($item != strip_tags($item)) {
+				throw new \App\Exceptions\Security('ERR_ILLEGAL_FIELD_VALUE||' . $this->getFieldModel()->getFieldName() . '||' . $this->getFieldModel()->getModuleName() . '||' . $item, 406);
+			}
+			if (!preg_match('/(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\.)+[a-z0-9][a-z0-9-]{0,61}[a-z0-9]/ui', $item)) {
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/uitypes/MultiDomain.php]: FILTER_VALIDATE_DOMAIN

File: modules/OSSMail/models/Mail.php (hunk 11 above)
Comment (rafalpospiech) [reply modules/OSSMail/models/Mail.php]: $moduleName is specified inside loop

File: modules/Vtiger/uitypes/MultiDomain.php [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+ * @copyright YetiForce Sp. z o.o.
+ * @license YetiForce Public License 3.0 (licenses/LicenseEN.txt or yetiforce.com)
+ * @author Rafal Pospiech <r.pospiech@yetiforce.com>
+ */
+class Vtiger_MultiDomain_UIType extends Vtiger_Base_UIType
+{
+	/**
+	 * {@inheritdoc}
+	 */
+	public function validate($value, $isUserFormat = false)
```
Comment (rskrzypczak) [inline modules/Vtiger/uitypes/MultiDomain.php]: too many conditions

File: layouts/basic/modules/Vtiger/Edit/Field/MultiDomain.tpl [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+{strip}
+<!-- tpl-Base-Edit-Field-MultiDomain -->
+	{assign var="FIELD_INFO" value=\App\Purifier::encodeHtml(\App\Json::encode($FIELD_MODEL->getFieldInfo()))}
+	{assign var="SPECIAL_VALIDATOR" value=$FIELD_MODEL->getValidator()}
+	{assign var=FIELD_VALUE value=$FIELD_MODEL->getEditViewDisplayValue($FIELD_MODEL->get('fieldvalue'),$RECORD)}
+<div class="tpl-Edit-Field-MultiDomain">
+	<input type="hidden" name="{$FIELD_MODEL->getFieldName()}" value="" />
+	<select id="{$MODULE}_{$VIEW}_fieldName_{$FIELD_MODEL->getName()}" title="{\App\Language::translate($FIELD_MODEL->getFieldLabel(), $MODULE)}" multiple data-tags="true" class="js-multi-domain select2 form-control col-md-12" name="{$FIELD_MODEL->getFieldName()}[]" data-fieldinfo='{$FIELD_INFO}' {if $FIELD_MODEL->isMandatory() eq true} data-validation-engine="validate[required,funcCall[Vtiger_Base_Validator_Js.invokeValidation]]" {else} data-validation-engine="validate[funcCall[Vtiger_Base_Validator_Js.invokeValidation]]"{/if} data-validator={\App\Json::encode([['name'=>'MultiDomain']])} {if $FIELD_MODEL->isEditableReadOnly()}readonly="readonly"{/if}>
```
Comment (rskrzypczak) [inline layouts/basic/modules/Vtiger/Edit/Field/MultiDomain.tpl]:
```suggestion
	<select id="{$MODULE_NAME}_{$VIEW}_fieldName_{$FIELD_MODEL->getName()}" title="{\App\Language::translate($FIELD_MODEL->getFieldLabel(), $MODULE_NAME)}" multiple data-tags="true" class="js-multi-domain select2 form-control col-md-12" name="{$FIELD_MODEL->getFieldName()}[]" data-fieldinfo='{$FIELD_INFO}' {if $FIELD_MODEL->isMandatory() eq true} data-validation-engine="validate[required,funcCall[Vtiger_Base_Validator_Js.invokeValidation]]" {else} data-validation-engine="validate[funcCall[Vtiger_Base_Validator_Js.invokeValidation]]"{/if} data-validator={\App\Json::encode([['name'=>'MultiDomain']])} {if $FIELD_MODEL->isEditableReadOnly()}readonly="readonly"{/if}>
```

File: app/Fields/MultiDomain.php [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+	{
+		$crmids = [];
+		$fieldName = $fieldModel->getName();
+		$moduleName = $fieldModel->getModuleName();
+		$queryGenerator = new \App\QueryGenerator($moduleName);
+		$queryGenerator->permissions = false;
+		if ($queryGenerator->getModuleField($fieldName)) {
+			$queryGenerator->setFields(['id']);
+			$queryGenerator->addNativeCondition(['like', $fieldName, ",$domain,"]);
+			$dataReader = $queryGenerator->createQuery()->createCommand()->query();
```
Comment (mariuszkrzaczkowski) [inline app/Fields/MultiDomain.php]: $crmids= $queryGenerator->createQuery()->column();

File: modules/OSSMail/models/Mail.php [hunk 22]
```diff
@@ -253,65 +252,115 @@ public function getEmail($name)
...
+		$crmids = [];
+		foreach ($emails as $email) {
+			$domain = mb_strtolower(explode('@', $email)[1]);
+			if (App\Cache::staticHas($cacheKey, $domain)) {
+				$cache = App\Cache::staticGet($cacheKey, $domain);
+				if ($cache != 0) {
+					$crmids = array_merge($crmids, $cache);
+				}
+			} else {
+				$crmids = App\Fields\MultiDomain::findIdByDomain($domain, $fieldModel);
```
Comment (mariuszkrzaczkowski) [inline modules/OSSMail/models/Mail.php]: ?

File: modules/OSSMail/models/Mail.php [hunk 23]
```diff
@@ -253,65 +252,115 @@ public function getEmail($name)
...
+	 * @return array crmids
+	 */
+	public function searchByEmails(array $emails, string $moduleName, string $fieldName)
+	{
+		$return = [];
+		foreach ($emails as $email) {
+			if (empty($email)) {
+				continue;
+			}
+			$cacheKey = 'MailSearchByEmails' . $moduleName . '_' . $fieldName;
```
Comment (mariuszkrzaczkowski) [inline modules/OSSMail/models/Mail.php]: move before the loop

File: modules/OSSMail/models/Mail.php [hunk 24]
```diff
@@ -253,65 +252,115 @@ public function getEmail($name)
...
+				if ($cache != 0) {
+					$return = array_merge($return, $cache);
+				}
+			} else {
+				$ids = [];
+				$queryGenerator = new \App\QueryGenerator($moduleName);
+				if ($queryGenerator->getModuleField($fieldName)) {
+					$queryGenerator->setFields(['id']);
+					$queryGenerator->addCondition($fieldName, $email, 'e');
+					$dataReader = $queryGenerator->createQuery()->createCommand()->query();
```
Comment (mariuszkrzaczkowski) [inline modules/OSSMail/models/Mail.php]:
```suggestion
					$ids= $queryGenerator->createQuery()->column();
```

File: modules/Vtiger/uitypes/MultiDomain.php [hunk 25]
```diff
@@ -0,0 +1,123 @@
...
+		}
+		return \App\Purifier::encodeHtml($value);
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function getEditViewDisplayValue($value, $recordModel = false)
+	{
+		return array_filter(explode(',', \App\Purifier::encodeHtml($value)));
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/uitypes/MultiDomain.php]: ?

File: modules/OSSMail/models/Mail.php [hunk 26]
```diff
@@ -253,65 +252,115 @@ public function getEmail($name)
...
+	 * @return array crmids
+	 */
+	public function searchByDomains(Vtiger_Field_Model $fieldModel, array $emails)
+	{
+		$fieldName = $fieldModel->getName();
+		$moduleName = $fieldModel->getModuleName();
+		$cacheKey = 'MailSearchByDomains' . $moduleName . '_' . $fieldName;
+		$crmids = [];
+		foreach ($emails as $email) {
+			$domain = mb_strtolower(explode('@', $email)[1]);
```
Comment (mariuszkrzaczkowski) [inline modules/OSSMail/models/Mail.php]: ?

**Pull request merged**